cloud.ganttproject.colloboque/src/main/kotlin/cloud/ganttproject/colloboque/DevServer.kt

```diff
@@ -73,13 +73,36 @@ class ColloboqueHttpServer(port: Int, private val colloboqueServer: ColloboqueSe
     }
 }
 
-class ColloboqueWebSocketServer(port: Int, private val colloboqueServer: ColloboqueServer) :
+class ColloboqueWebSocketServer(port: Int, private val colloboqueServer: ColloboqueServer,
+                                private val updateInputChannel: Channel<InputXlog>,
+                                private val serverResponseChannel: Channel<String>) :
   NanoWSD("localhost", port) {
-  override fun openWebSocket(handshake: IHTTPSession?): WebSocket {
-    return WebSocketImpl(handshake)
+  override fun openWebSocket(handshake: IHTTPSession): WebSocket {
+    return WebSocketImpl(handshake, colloboqueServer, updateInputChannel, serverResponseChannel)
   }
 
-  private class WebSocketImpl(handshake: IHTTPSession?) : WebSocket(handshake) {
+  @OptIn(DelicateCoroutinesApi::class)
+  private class WebSocketImpl(handshake: IHTTPSession,
+                              private val colloboqueServer: ColloboqueServer,
+                              private val updateInputChannel: Channel<InputXlog>,
+                              private val serverResponseChannel: Channel<String>) : WebSocket(handshake) {
+    init {
+      handshake.parameters["projectRefid"]?.firstOrNull()?.also { refid ->
+        colloboqueServer.init(refid, false)
+        this.handshakeResponse.addHeader("baseTxnId", colloboqueServer.getBaseTxnId(refid))
+      }
+      GlobalScope.launch {
```

Comment:
I suggest creating an individual scope for sending the responses (and actually for other tasks as well, as it provides more control), e.g. like this:

```
private val responseScope = CoroutineScope(Executors.newSingleThreadExecutor().asCoroutineDispatcher())
...
responseScope.launch {}
```

Reply:
Do you mean having an individual scope per instance (rather then GlobalScope) or per each need (sending/receiving)?

Reply:
individual scope for each activity, one for receiving, one for sending.